bundles/org.openhab.core.model.yaml/src/main/java/org/openhab/core/model/yaml/internal/rules/YamlRuleDTO.java

```diff
@@ -0,0 +1,396 @@
+
+    @Override
+    public YamlElement cloneWithoutId() {
+        YamlRuleDTO copy;
+        try {
+            copy = (YamlRuleDTO) super.clone();
+            copy.uid = null;
+            return copy;
+        } catch (CloneNotSupportedException e) {
+            // Will never happen
+            return new YamlRuleDTO();
+        }
+    }
+
+    @Override
+    public boolean isValid(@Nullable List<@NonNull String> errors, @Nullable List<@NonNull String> warnings) {
+        // Check that uid is present
+        if (uid == null || uid.isBlank()) {
+            addToList(errors, "invalid rule: uid is missing");
+            return false;
+        }
+        boolean ok = true;
+        // Check that uid only contains valid characters
+        String[] segments = uid.split(AbstractUID.SEPARATOR);
+        for (String segment : segments) {
+            if (!UID_SEGMENT_PATTERN.matcher(segment).matches()) {
+                addToList(errors, "invalid rule \"%s\": segment \"%s\" in the uid doesn't match the expected syntax %s"
+                        .formatted(uid, segment, UID_SEGMENT_PATTERN.pattern()));
+                ok = false;
+            }
+        }
+
+        // Check that name is present
+        if (label == null || label.isBlank()) {
+            addToList(errors, "invalid rule \"%s\": label is missing".formatted(uid));
+            ok = false;
+        }
+
+        // Check that the rule either has configuration (rule stub) or that it has at least one module
+        if ((config == null || config.isEmpty()) && (triggers == null || triggers.isEmpty())
+                && (conditions == null || conditions.isEmpty()) && (actions == null || actions.isEmpty())) {
+            addToList(errors, "invalid rule \"%s\": the rule is empty".formatted(uid));
+            ok = false;
+        }
+
+        // Check that module IDs are unique
+        Set<String> ids = new HashSet<>();
+        ok &= enumerateModuleIds(triggers, ids, errors);
+        ok &= enumerateModuleIds(conditions, ids, errors);
+        ok &= enumerateModuleIds(actions, ids, errors);
+
+        return ok;
+    }
+
+    private boolean enumerateModuleIds(@Nullable List<@NonNull ? extends YamlModuleDTO> modules,
+            @NonNull Set<String> ids, @Nullable List<@NonNull String> errors) {
+        if (modules == null) {
+            return true;
+        }
+        String id;
+        for (YamlModuleDTO module : modules) {
+            if ((id = module.id) == null || id.isBlank()) {
+                continue;
+            }
+            if (ids.contains(id)) {
+                String moduleType;
+                if (module instanceof YamlActionDTO) {
+                    moduleType = "action";
+                } else if (module instanceof YamlConditionDTO) {
+                    moduleType = "condition";
+                } else {
+                    moduleType = "trigger";
+                }
+                addToList(errors, "illegal " + moduleType + " ID '" + id
```

Comment:
Please start your error message with: "invalid rule "xxx": " so that the user will know in which rule the problem is.
You have the choice to add it in the calling method, that is `isValid`

Reply:
Same remark is applicable for YamlRuleTemplateDTO

Reply:
Done